macros/timeHistoryAnalyzer.py

```diff
@@ -0,0 +1,457 @@
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.maskReason[vfat,channel] != 0,
+                             maxSkip)
+
+    return list(filter(lambda r: r.badMaskReasonScanCount() >= minBadScans,
+                       ranges))
+
+def findRangesMask(data, vfat, channel, maxSkip = 5, minBadScans = 4):
+    """Finds ranges of scans based on the mask attribute.
+
+    Searches the data for the given vfat and channel for ranges of scans with
+    non-zero maskReason. During the search, at most maxSkip scans with mask not
+    set can be skipped. Only ranges with more than minBadScans are kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        maxSkip: The maximum number of "good" scans between two "bad" scans
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.mask[vfat,channel] != 0,
+                             maxSkip)
+
+    return list(filter(lambda r: r.maskedScanCount() >= minBadScans,
+                       ranges))
+
+def findRangesNoise(data,
+                    vfat,
+                    channel,
+                    threshold = 0.02,
+                    maxSkip = 5,
+                    minBadScans = 4):
+    """Finds ranges of scans based on noise.
+
+    Searches the data for the given vfat and channel for ranges of scans with
+    noise < threshold. During the search, at most maxSkip scans with mask not
+    set can be skipped. Only ranges with more than minBadScans are kept.
+
+    Args:
+        data: The TimeSeriesData object to pull data from
+        vfat: The VFAT to return ranges for
+        channel: The channel to return ranges for
+        threshold: The minimum value of noise for a scan to be "good"
+        maxSkip: The maximum number of "good" scans between two "bad" scans
+        minBadScans: The minimum number of bad scans
+
+    Returns:
+        A list of MaskedRange objects
+    """
+    ranges = _findRangesMeta(data,
+                             vfat,
+                             channel,
+                             data.noise[vfat,channel] < threshold,
+                             maxSkip)
+
+    return list(filter(lambda r: np.count_nonzero(r.noise() < threshold) >= minBadScans,
+                       ranges))
+
+class TimeSeriesData(object):
+    """Holds information about time variation of scan results.
+
+    Each property is stored as a 3D Numpy array with indexes
+    [vfat][strip][time]. The time index -> scan date mapping is exposed in the
```

Comment:
in all cases change `strip` to `chanStripOrPanPin` since this could represent three different quantities and is not always "strip"